manual/runtime-files.wiki

```diff
@@ -0,0 +1,112 @@
+JavaScript runtime files for some OCaml libraries used by the
+executable. Tracking the required runtime files is done using the de-facto
+standard for library management in OCaml, Findlib.
+
+== The old deprecated way
+
+Before dune and its native support for js_of_ocaml, the (hacky) way
+to rely on Findlib was to use a **linkopts** variable with a **javascript**
+predicate {{{linkopts(javascript)}}} and have js_of_ocaml be responsible for
+looking up the files in the right directory.
+
+This approach is not very satisfying:
+* It makes it harder for a build system to know the set of dependencies.
+* The js_of_ocaml compiler needs to perform findlib resolution itself.
+
+=== Listing linkopts
+
+To list all {{{linkopts(javascript)}}} for a set of libraries:
+
+{{{
+  ocamlfind query -o-format -r -predicates javascript LIB1 LIB2 ...
+}}}
+
+For example, for **base** and **ppx_expect**
+
+{{{
+  $ ocamlfind query -o-format -r -predicates javascript base ppx_expect
+  +base/base_internalhash_types/runtime.js
+  +base/runtime.js
+  +time_now/runtime.js
+  +ppx_expect/collector/runtime.js
+}}}
+
+=== Declaring runtime JavaScript files inside META files
+
+The content of {{{linkopts(javascript)}}} is given directly to the
+js_of_ocaml compiler.  To be able reference files independently of the
+installation path, we rely on js_of_ocaml's findlib support (see Command-Line findlib support below).
+
+For example, to declare that **mypackage.sublib** requires the file **mypackage/subdir/runtime.js**
+
+{{{
+  cat mypackage/META
+  package "sublib" (
+    directory = "subdir"
+    linkopts(javascript) = "+mypackage/sublib/runtime.js"
+  )
+}}}
+
+
+== The new way (introduced by/for dune)
+
+Using findlib more efficiently, with a usage closer to what's
+done for OCaml archives.
+
+=== Listing js runtime files
+
+To list all JavaScript runtime files for a set of libraries:
+
+{{{
+  ocamlfind query -format "%+(jsoo_runtime)" -r LIB1 LIB2 ... | grep -v "^$"
+}}}
+
+For example, for **base** and **ppx_expect**
+
+{{{
+  $ ocamlfind query -format "%+(jsoo_runtime)" -r base ppx_expect | grep -v "^$"
+  ~/.opam/4.14.0/lib/base/base_internalhash_types/runtime.js
+  ~/.opam/4.14.0/lib/base/runtime.js
+  ~/.opam/4.14.0/lib/time_now/runtime.js
+  ~/.opam/4.14.0/lib/ppx_expect/collector/runtime.js
+}}}
+
+=== Declaring runtime JavaScript files inside META files
+
+The **jsoo_runtime** variable can be used to declare one or more runtime files.
+Files must be coma-separated and should be plain names relative to the package directory.
+
+To declare that **mypackage.sublib** requires the file **mypackage/subdir/runtime.js**
+{{{
+  cat mypackage/META
+  package "sublib" (
+    directory = "subdir"
+    jsoo_runtime = "runtime.js"
+  )
+}}}
+
+== Command line findlib support
+
+If js_of_ocaml is compiled with findlib support, one can use the
+following syntax to reference JavaScript files:
+* {{{ +{findlib-package}/{javascript-file} }}}
+* {{{ +{javascript-file} }}}
```

Comment:
I'm a bit uneasy about that convention, is it here for backward compat or is that new ? If it's new I would rather say let's not add it or use a different symbol for packages. If it's already there, I would suggest to deprecate it.

IIRC in the `ocaml` compilers the `+` simply means substitute `ocamlc -where`. So `+dir/file` will be `$(ocamlc -where)/dir/file` not file `file` in package `dir`.

Reply:
This is not new, this has been present for many years. It was introduced before there was any build system support for js_of_ocaml.

Findlib has `@package/file` which I should probably have used back then.

> The variable "archive" specifies the list of archive files. These files should be given either as (1) plain names without any directory information; they are only searched in the package directory. (2) Or they have the form "+path" in which case the files are looked up relative to the standard library. (3) Or they have the form "@name/file" in which case the files are looked up in the package directory of another package. (4) Or they are given as absolute paths.
```

Reply:
Dropping special findlib support seems to be the way to go.

Reply:
The documentation now mentions that findlib support is deprecated.